datafusion/expr/src/expr_schema.rs

```diff
@@ -158,7 +160,29 @@ impl ExprSchemable for Expr {
                     .iter()
                     .map(|e| e.get_type(schema))
                     .collect::<Result<Vec<_>>>()?;
-                fun.return_type(&data_types)
+                match fun {
+                    WindowFunctionDefinition::AggregateUDF(udf) => {
+                        let new_types = data_types_with_aggregate_udf(&data_types, udf).map_err(|err| {
+                            plan_datafusion_err!(
+                                "{} and {}",
+                                err,
+                                utils::generate_signature_error_msg(
+                                    fun.name(),
+                                    fun.signature().clone(),
+                                    &data_types
+                                )
+                            )
+                        })?;
+                        Ok(fun.return_type(&new_types)?)
+                    }
```

Comment:
Maybe we should bury this check, and conversion inside to the `fun.return_type` implementation for `WindowFunctionDefinition::AggregateUDF` not sure though.

Reply:
I prefer separate `coerce_types` and `return_types`.